packages/runner-ct/src/iframe/iframes.tsx

```diff
@@ -22,134 +21,135 @@ export function getSpecUrl ({ namespace, spec }, prefix = '') {
+      highlightEl: autIframe.highlightEl,
+      detachDom: autIframe.detachDom,
+      snapshotControls: (snapshotProps) => (
+        <SnapshotControls
+          eventManager={props.eventManager}
+          snapshotProps={snapshotProps}
+          state={props.state}
+          onToggleHighlights={_toggleSnapshotHighlights}
+          onStateChange={_changeSnapshotState}
         />
-        <ScriptError error={scriptError} />
-        <div className='cover' />
-      </div>
-    )
-  }
-
-  componentDidMount () {
-    const config = this.props.config
-
-    this.autIframe = new AutIframe(config)
+      ),
+    }),
+  )
 
-    this.props.eventManager.on('visit:failed', this.autIframe.showVisitFailure)
-    this.props.eventManager.on('before:screenshot', this.autIframe.beforeScreenshot)
-    this.props.eventManager.on('after:screenshot', this.autIframe.afterScreenshot)
-    this.props.eventManager.on('script:error', this._setScriptError)
+  useEffect(() => {
+    props.eventManager.on('visit:failed', autIframe.showVisitFailure)
+    props.eventManager.on('before:screenshot', autIframe.beforeScreenshot)
+    props.eventManager.on('after:screenshot', autIframe.afterScreenshot)
+    props.eventManager.on('script:error', _setScriptError)
 
     // TODO: need to take headless mode into account
     // may need to not display reporter if more than 200 tests
-    this.props.eventManager.on('restart', () => {
-      this._run(this.props.state.spec, config)
+    props.eventManager.on('restart', () => {
+      _run(props.state.spec, props.config)
     })
 
-    this.props.eventManager.on('print:selector:elements:to:console', this._printSelectorElementsToConsole)
+    props.eventManager.on('print:selector:elements:to:console', _printSelectorElementsToConsole)
+
+    props.eventManager.start(props.config)
+
+    setDisposers([
+      autorun(() => {
+        autIframe.toggleSelectorPlayground(selectorPlaygroundModel.isEnabled)
+      }),
+      autorun(() => {
+        autIframe.toggleSelectorHighlight(selectorPlaygroundModel.isShowingHighlight)
+      }),
+      autorun(() => {
+        if (props.state.spec) {
+          _run(props.state.spec, props.config)
+        }
+      }),
+    ])
+
+    iframeModel.listen()
+
+    return () => {
+      props.eventManager.notifyRunningSpec(null)
+      props.eventManager.stop()
+      disposers.forEach((dispose) => {
+        dispose()
+      })
+    }
+  }, [])
 
-    this._disposers.push(autorun(() => {
-      this.autIframe.toggleSelectorPlayground(selectorPlaygroundModel.isEnabled)
-    }))
+  useEffect(() => {
+    props.state.callbackAfterUpdate?.()
+  }, [props.state.callbackAfterUpdate])
 
-    this._disposers.push(autorun(() => {
-      this.autIframe.toggleSelectorHighlight(selectorPlaygroundModel.isShowingHighlight)
-    }))
+  const _toggleSnapshotHighlights = (snapshotProps) => {
+    props.state.snapshot.showingHighlights = !props.state.snapshot.showingHighlights
```

Comment:
Obviously mutation `props.state` is not very good - I tried to make this refactor without changing the logic at all. I'll fix this in another pass in the future.

Reply:
I expect this is actually correct. `state` is likely a MobX store, so there's nothing wrong with mutating it. That being said, it's much nicer to write methods on the store to perform the mutation, rather than doing it yourself inline.

Reply:
I see. Technically correct or not, I agree with you, it *looks* weird and just seems unidiomatic. I'll make a method to handle the mutation.

Reply:
Fixed.